web/packages/teleport/src/Discover/SelectResource/SelectResource.tsx

```diff
@@ -418,6 +411,52 @@ export function sortResources(
   return sortedResources;
 }
 
+/**
+ * Returns prioritized resources based on user preferences cluster state
+ *
+ * @remarks
+ * A user can have preferredResources set via onboarding either from the survey (preferredResources)
+ * or various query parameters (marketingParams). We sort the list by the marketingParam if available.
+ * If not, we sort by preferred resource type if available.
+ * We do not search.
+ *
+ * @param preferences - Cluster state user preferences
+ * @returns PrioritizedResources which is both the resource to prioritize and a boolean value of the value
+ *
+ */
+function getPrioritizedResources(
+  preferences: UserPreferences
+): PrioritizedResources {
+  const marketingParams =
+    preferences && preferences.onboard && preferences.onboard.marketingParams;
+
+  if (marketingParams) {
+    const marketingPriorities = getMarketingMatches(marketingParams);
+    if (marketingPriorities.length > 0) {
+      return {
+        hasPreferredResources: true,
+        preferredResources: marketingPriorities,
+      };
+    }
+  }
+
+  const preferredResources =
+    (preferences &&
+      preferences.onboard &&
+      preferences.onboard.preferredResources) ||
+    [];
```

Comment:
I think we can simplify it as well:
```suggestion
  const preferredResources: preferences.onboard?.preferredResources || [];
```